src/Microsoft.AspNetCore.Blazor.Browser/Interop/RegisteredFunction.cs

```diff
@@ -24,9 +26,53 @@ public static TRes Invoke<TRes>(string identifier, params object[] args)
             // This is a low-perf convenience method that bypasses the need to deal with
             // .NET memory and data structures on the JS side
             var argsJson = args.Select(JsonUtil.Serialize);
+
             var resultJson = InvokeUnmarshalled<string>("invokeWithJsonMarshalling",
                 argsJson.Prepend(identifier).ToArray());
-            return JsonUtil.Deserialize<TRes>(resultJson);
+
+            var result = JsonUtil.Deserialize<InvocationResult<TRes>>(resultJson);
+            if (result.Succeeded)
+            {
+                return result.Result;
+            }
+            else
+            {
+                throw new JavaScriptException(result.Message);
+            }
+        }
+
+        /// <summary>
+        /// Invokes the JavaScript function registered with the specified identifier.
+        /// Arguments and return values are marshalled via JSON serialization.
+        /// </summary>
+        /// <typeparam name="TRes">The .NET type corresponding to the function's return value type. This type must be JSON deserializable.</typeparam>
+        /// <param name="identifier">The identifier used when registering the target function.</param>
+        /// <param name="args">The arguments to pass, each of which must be JSON serializable.</param>
+        /// <returns>The result of the function invocation.</returns>
+        public static Task<TRes> InvokeAsync<TRes>(string identifier, params object[] args)
+        {
+            var tcs = new TaskCompletionSource<TRes>();
+            var argsJson = args.Select(JsonUtil.Serialize);
+            var callbackId = Guid.NewGuid().ToString();
+
+            TrackedReference.Track(callbackId, new Action<string>(r =>
+            {
+                var res = JsonUtil.Deserialize<InvocationResult<TRes>>(r);
+                if (res.Succeeded)
+                {
+                    tcs.SetResult(res.Result);
+                }
+                else
+                {
+                    tcs.SetException(new JavaScriptException(res.Message));
+                }
+            }));
+
+            var result = Invoke<TRes>(
+                    "invokeWithJsonMarshallingAsync",
+                    new[] { identifier, callbackId }.Concat(argsJson).ToArray());
```

Comment:
The code in this method is pretty allocatey for such a low-level primitive.

* Instead of instantiating and tracking the `Action`, could we track just the `TaskCompletionSource`?
* When building the array of args on line 73, could we construct an array of length `argsJson.Length + 2` and then use Array.Copy to put `argsJson` in starting at index 2? Otherwise we're allocating multiple temporary objects along the way to building this.